### PR TITLE
Integration tests to check number of vcpus and available memory from within the guest

### PR DESCRIPTION
**Pull request by dodsonmg**
Addresses issues #26 and #27.

### CODE REVIEW

File: tests/test_run_reference_vmm.py [hunk 1]
```diff
@@ -156,3 +180,22 @@ def test_reference_vmm_num_vcpus(kernel):
         expect_vcpus(vmm_process, expected_vcpus)
 
         shutdown(vmm_process)
+
+@pytest.mark.parametrize("kernel", KERNELS_INITRAMFS)
+def test_reference_vmm_mem(kernel):
+    """Start the reference VMM and verify the amount of memory."""
+
+    mem_sizes_mib = [512, 1024, 2048]
```
Comment (andreeaflorescu) [inline tests/test_run_reference_vmm.py:188]: Can you check with a value that is higher that the start of the MMIO Gap address? The value is defined here: https://github.com/rust-vmm/vmm-reference/blob/1c5ed93058ba83b87a18052a8ef2975752932edd/src/vmm/src/lib.rs#L60)

We are creating 2 memory regions when the memory does not fit before the MMIO gap, and we should probably check that we still get the expected amount of memory in that situation.

File: tests/test_run_reference_vmm.py [hunk 2]
```diff
@@ -102,6 +102,21 @@ def expect_string(vmm_process, expected_string):
         if expected_string in nextline.decode():
             break
 
+def expect_vcpus(vmm_process, expected_vcpus):
+    # /proc/cpuinfo displays info about each vCPU
+    vmm_process.stdin.write(b'cat /proc/cpuinfo\n')
+    vmm_process.stdin.flush()
+
+    # The sibling field for each vCPU displays the total number of vCPUs
+    expected_siblings = "siblings:{}".format(expected_vcpus)
```
[the remark this answers is not in the document]
Comment (dodsonmg) [reply tests/test_run_reference_vmm.py:105]: Yes.  That makes sense.  Will fixup.

File: tests/test_run_reference_vmm.py (hunk 1 above)
Comment (dodsonmg) [reply tests/test_run_reference_vmm.py:188]: If I'm reading it right, I need a value > 4GB?

That makes sense.  I can also explicitly test the edge cases.  Will fixup.

Comment (andreeaflorescu) [reply tests/test_run_reference_vmm.py:188]: Yap, that's correct. We can remove some other values from the `mem_sizes_mib` list so that the test does not take too long.

Comment (dodsonmg) [reply tests/test_run_reference_vmm.py]: Ok.  I'm now testing well below, well above, and then right at the start of the MMIO gap to verify that the partitioning logic in `create_guest_memory()` is working:

https://github.com/rust-vmm/vmm-reference/blob/1c5ed93058ba83b87a18052a8ef2975752932edd/src/vmm/src/lib.rs#L222-L230

If that's overkill, I can pull some of it out.

Comment (andreeaflorescu) [reply tests/test_run_reference_vmm.py]: I think that's fine for now. In the future we can have some more optimization so the tests take less (like for example grouping together the tests for memory size, vcpu count, and maybe disk?), but for now it's not necessary as we're also still figuring out a testing framework.

**Review by andreeaflorescu — APPROVED** [review]

File: tests/test_run_reference_vmm.py [hunk 3]
```diff
@@ -102,6 +96,52 @@ def expect_string(vmm_process, expected_string):
...
+            nextline = "".join(nextline.split())
+            # extract the number of vCPUs
+            actual_vcpus = int(nextline.split(':')[1])
+            break
+
+    assert actual_vcpus == expected_vcpus, \
+            "Expected {}, found {} vCPUs".format(expected_vcpus, actual_vcpus)
+
+def expect_mem(vmm_process, expected_mem_mib):
+    expected_mem_kib = expected_mem_mib * 1000
```
Comment (lauralt) [inline tests/test_run_reference_vmm.py:120]: Shouldn't 1000 be 1024?

File: tests/test_run_reference_vmm.py [hunk 4]
```diff
@@ -102,6 +96,52 @@ def expect_string(vmm_process, expected_string):
...
+        nextline = vmm_process.stdout.readline().decode()
+        if "MemTotal" in nextline:
+            break
+
+    # Remove whitespace and extract the memory value
+    # Expected format is 'MemTotal: [value] kB'
+    mem_total = nextline.split()[1]
+
+    # Expect the difference between the expected and actual memory
+    # to be less than 10%
```
Comment (andreeaflorescu) [inline tests/test_run_reference_vmm.py:140]: On a second thought, 10% seems like a lot. Would a fixed diff in KB work as well?

**Comment by dodsonmg** [comment]
I hadn't noticed the recent commits to `master`, so I just pulled them in, rebased, and force-pushed, but the CI failed almost immediately.

I don't seem to be able to see the test results (other than a big, red circle-X).  The tests are passing for me locally, though, so if someone could give me a hint, I would be grateful.

**Comment by andreeaflorescu** [comment]
> I hadn't noticed the recent commits to `master`, so I just pulled them in, rebased, and force-pushed, but the CI failed almost immediately.
> 
> I don't seem to be able to see the test results (other than a big, red circle-X). The tests are passing for me locally, though, so if someone could give me a hint, I would be grateful.

@dodsonmg we had some issues with the CI, but they should be fixed now. The tests in this PR are still failing, but it seems that it is a change related to this PR. Can you remove the rust-vmm-ci update from your first commit? I have a hunch that might be the problem.

**Comment by lauralt** [comment]
Besides Andreea's comment, changes look good to me :+1:.

**Review by lauralt — APPROVED** [review]

**Comment by dodsonmg** [comment]
Whew... Sorry that took so long.  Thanks for helping me through!

**Review by andreeaflorescu — APPROVED** [review]
